test/fact_importer_test_helper.rb

```diff
@@ -6,32 +6,16 @@ def allow_transactions_for(importer)
     importer.stubs(:ensure_no_active_transaction).returns(true)
   end
 
-  module ClassMethods
-    def allow_transactions_for_any_importer
-      FactImporter.singleton_class.prepend FactImporterFactoryStubber
-
-      FactImporter.register_instance_stubs do |importer_class|
-        importer_class.any_instance.stubs(:ensure_no_active_transaction).returns(true)
-      end
+  def allow_transactions_for_any_importer
+    FactImporter.importers.values.each do |importer|
+      allow_transactions_for(importer.any_instance)
     end
   end
-end
-
-module FactImporterFactoryStubber
-  def register_instance_stubs(&block)
-    instance_stubs << block
-  end
 
-  def importer_for(*args)
-    instance = super
-    instance_stubs.each do |stub_block|
-      stub_block.call(instance)
+  module ClassMethods
+    def allow_transactions_for_any_importer
+      setup :allow_transactions_for_any_importer
     end
-    instance
-  end
-
-  def instance_stubs
-    @instance_stubs ||= []
   end
 end
```

Comment:
Are both instance and class methods used? Wondering if we can drop one.

Reply:
Class method is the one that is "customer facing" - the one that would be declared by the user. Instance method is the actual implementation, sorry didn't have enough imagination to call it differently. The instance method would be called by the test framework (since we register it as setup callback) and it contains the actual isolation code.

The definition would look like this:

``` ruby
class HostTest < ActiveSupport::TestCase
  include FactImporterIsolation

  allow_transactions_for_any_importer
  # ^ Usage of the class method
```